src/js/background/context-menu.js

```diff
@@ -0,0 +1,388 @@
+      
+        // Note: If user is already premium, this will return false.
+        return !premium;
+      },
+    },
+    onCreatedCallback: ()=> {
+      // Catch errors when trying to create the same menu twice.
+      // The browser.menus API is limited. You cannot query if a menu item already exists.
+      // The error it throws does not show up to the user.
+
+      if (browser.runtime.lastError) {
+        return;
+      }
+    }
+  }
+};
+
+// TLDR: We make sure the menu is still open when performing the refresh command, as to not waste resoruces. 
+// This method is taken from https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/menus/onShown documentation. 
+let lastMenuInstanceId = 0;
+let nextMenuInstanceId = 1;
+
+// Events
+browser.menus.onShown.addListener(async (info, tab) => {
+  
+  if (!info.menuIds.includes("fx-private-relay-generate-alias") ) {
+    // No Relay menu items exist. Stop listening.
+    return;
+  }
+
+  let menuInstanceId = nextMenuInstanceId++;
+  lastMenuInstanceId = menuInstanceId;
+
+  const domain = relayContextMenus.utils.getHostnameFromUrlConstructor(tab.url);
+  await relayContextMenus.init(domain);
+
+  if (menuInstanceId !== lastMenuInstanceId) {
+    return; // Menu was closed and shown again.
+  }
+});
+
+browser.menus.onHidden.addListener(async (info, tab) => {
+  lastMenuInstanceId = 0;
+});
+
+browser.menus.onClicked.addListener(async (info, tab) => {
+  switch (info.menuItemId) {
+    case "fx-private-relay-generate-alias":
+      sendMetricsEvent({
+        category: "Extension: Context Menu",
+        action: "click",
+        label: "context-menu-generate-alias",
+      });
+      await relayContextMenus.listeners.onMakeRelayAddressForTargetElement(info, tab);
+      break;
+    case "fx-private-relay-get-unlimited-aliases":
+      sendMetricsEvent({
+        category: "Extension: Context Menu",
+        action: "click",
+        label: "context-menu-get-unlimited-aliases",
+      });
+      const { fxaSubscriptionsUrl, premiumProdId, premiumPriceId } =
+        await browser.storage.local.get([
+          "fxaSubscriptionsUrl",
+          "premiumProdId",
+          "premiumPriceId",
+        ]);
+      const urlPremium = `${fxaSubscriptionsUrl}/products/${premiumProdId}?plan=${premiumPriceId}`;
+      await browser.tabs.create({ url: urlPremium });
+      break;
+    case "fx-private-relay-manage-aliases":
+      sendMetricsEvent({
+        category: "Extension: Context Menu",
+        action: "click",
+        label: "context-menu-relay-manage-aliases",
+      });
+      const urlManageAliases = `${RELAY_SITE_ORIGIN}/accounts/profile/`;
+      await browser.tabs.create({ url: urlManageAliases });
+      break;
+  }
+
+  if (info.menuItemId.startsWith("fx-private-relay-use-existing-alias_")) {
+    await relayContextMenus.listeners.onFillInAddressWithAliasId(info, tab);
+  }
+
+});
+
+(async () => {
+  await relayContextMenus.init();
+})();
```

Comment:
As in the other file, does this need the wrapping?

```suggestion
relayContextMenus.init();
```

Reply:
~Because it's an async function, and needs to be called in an [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) way. When `relayContextMenus.init();` runs (without the `await`), we get an error.~

Yeah, this works great! :)

Reply:
TIL You don't always have to add `await` to `async` functions.

Reply:
Haha :grin: The thing to keep in mind is that a function marked as `async` just returns a Promise behind the scenes, and `await`ing it is just a way to run whatever comes after in that Promise's `.then`.

Reply:
Removed in 787f56125a7bdf2aa0bd51eb60f64daecf080440